### DISCUSSION TITLE
add search possibility to sidebar

### DISCUSSION THREAD

**Discussion by ryg-** [neomutt/neomutt#4499] [Development]
I would like to have possibility to quickly search mailbox in sidebar.

E.g., as usually,  print few characters of mailbox after pressing search hot key (switch to search mode)

After entering few characters, probably, I should see filtered list of mailboxes and on that level I should be able 
select mailbox which I want and open it.

As i know, there no possibility to simulate that kind of behavior via hooks, scripts and so on.

**Comment by flatcap** [comment]
Years ago, I knocked up a quick prototype for this...

[![asciicast](https://asciinema.org/a/82605.svg)](https://asciinema.org/a/82605)

Code in the [[matcher]](https://github.com/neomutt/neomutt-old/tree/matcher) branch, in the [neomutt-old](https://github.com/neomutt/neomutt-old#neomutt-2017-10-13) repo

... but I didn't have time to work out all the details.

**Would anyone like to volunteer?**  (help will be given)

**Comment by lucilanga** [comment]
Hi flatcap,
as I understand the [matcher] functionality works like a quick search. You have to enter a few letters and you'll be shown a filtered list of mailboxes. If it's an exact match pressing enter will get you to that mailbox, it it's not an exact search pressing enter will get you to the first entry in the filtered list.

Now I understand (or at least I claim to understand) why it's done like this, but I do not think it is very muttesque way to do it and I really hate to bring mutt_enter_string() back, let alone put a Sb redraw callback on it.

To be perfectly honest most users won't even bother editing the query to get a perfect match, most of the time everybody wants to type 2-3 letters and go _up_ or _down_ from there.

The mutt way to do it, or at least the way I imagine this, is bringing the search/limit concept that we have for index into sidebar, which now after all refactoring done over time it should be easy to implement.

eg. For sidebar search it's just a matter of getting the query from buf_get_field()  then ARRAY_FOREACH ->strcasecmp-> wdata->hil_index finally WA_RECALC and Bob's your uncle.

For sidebar limit it might be a little more complicated than that but still it's just a matter of manipulating is_hidden flag perhaps adding a separate flags for limit purpose like it's done for index.

Nonetheless I would say you got a suc^H^H^H .. erm .. volunteer.

**Comment by flatcap** [comment]
> as I understand the [matcher] functionality works like a quick search.

yes

> most of the time everybody wants to type 2-3 letters and go up or down from there

That _was_ the plan, but it was just a quick hack.

> ... why it's done like this
> I really hate to bring mutt_enter_string() back
> let alone put a Sb redraw callback on it.

A **lot** of code has changed since then.
I'm intrigued now.  I'll have a think how I'd do it better, these days.

> The mutt way to do it ... search/limit concept

er, yes.  But that's solving a different problem.

The Matcher is a navigation tool.
Limiting is a change of view.
Unless you plan to limit pattern, select, limit all.

That's OK.  `<sidebar-limit>` is useful.

> ... it's just a matter of ...

Yes, that sounds about right

> volunteer

Great!

It sounds like you've been have a good read through the code.
Feel free to ask **lots** of questions.

If you want to discuss technical things, we could open a discussion:

- https://github.com/neomutt/neomutt/discussions

Or you can find me on IRC most days (European daylight hours, generally)

- `#neomutt` on `irc.libera.chat`

Thanks!

**Comment by lucilanga** [comment]
> The Matcher is a navigation tool.
> Limiting is a change of view.
> Unless you plan to limit pattern, select, limit all.

So I went and played a little bit more with the matcher, the thing is
really like the quick/search/type thing about it.
So I went ahead and put a callback to buf_get_field() as per your
original implementation, no worries there.
The thing kinda works, the way I see it there are two ways about it:

first we manipulate hil_index, the end result is the highlight cursor
jumping all over the place about as you type in the pattern (we do have
an issue about delete but that's something I'll able to tack later)

second we do this by temporary limiting the view, as per your original
implementation, we do this by resetting top/bot/hil indexes and
manipulate is_hidden/visible, the only thing I hate about this we lose
hil cursor, at least till we narrow the search to one result. Or to put
it differently presenting 3 results with only the first one selected and
no possibility to up/down the next one, well it kinda sucks. What do you
do then ? delete ? ^G it ? enter it ?

It might entirely be possible I'm overanalysing things but I prefer the
first one. It's really cheap to manipulate hil_index, the user already
knows his mailboxes names, the cursor might jump a bit erratic but still
workable I think. It's not like he's going to type in Tchaikovsky for a
mailbox :)

I could post a screencast but I only have my personal mailboxes right
now, I'll see what I can do about it. I think you mentioned somewhere
using a test dataset...but I do not remember where.

Now the less fun part, I've looked over the source and there are quite
a few places where we have buf_get_field() I could either update each
occurrence or do a _buf_get_field(), sort of what you did with
_mutt_enter_string, or even better do a trimmed down version like
buf_get_field_callback(), however I'm not entirely sure where we would
use such quick input field in the near future, nor I fancy maintaining
such a function over several refactoring jobs.

I know I said I'll do the limit thing but I got to think about it and
I'll probably end up overanalysing that as well and there are other things
I want to work on right now like like #1114.

I remember you said something about fixing different issue with
search/limit view so I might come around to fix that later.

nb. replying to github issue directly

**Comment by flatcap** [comment]
Sorry for the slow reply; I've been having a dig around in the code.

> the thing is I really like the quick/search/type thing about it

ooh, good :-)

> So I went ahead and put a callback to buf_get_field()
> The thing kinda works

I have some ideas how we _could_ do it differently (below).

> the way I see it there are two ways about it:

> first we manipulate hil_index, the end result is the highlight cursor
> jumping all over the place about as you type in the pattern

As the user types, the set of visible folders will shrink.

Do we anchor the view:
- At the beginning
  The highlight is dragged into the current view

- On the highlight
  The position of the highlight doesn't change until there are too few folders to fill the view

If the highlighted folder doesn't match, the highlight is moved to the next/previous folder.

> (we do have an issue about delete but that's something I'll able to tackle later)

When deleting chars, the set will increase in size.
If the view is anchored:
- At the beginning
  The highlighted folder is changed to keep it in the view

- On the highlight
  The position of the highlight doesn't change, but the contents of the view might

> second we do this by temporary limiting the view, as per your original implementation,

remember, quick hack :-)

> the only thing I hate about this we lose hil cursor,
> at least till we narrow the search to one result.

I think we need an overlay for the quick search... (maybe for Phase 2 :-)

If the user already has a limit in place, e.g. `sidebar_new_mail_only=yes`
then does the quick search only apply to those folders?

> presenting 3 results with only the first one selected and
> no possibility to up/down the next one, well it kinda sucks.

In the current `buf_get_field()`, <kbd>\<up\></kbd> and <kbd>\<down\></kbd> are mapped to the history:

```c
{ OP_EDITOR_HISTORY_DOWN, "<down>" },
{ OP_EDITOR_HISTORY_UP,   "<up>" },
```

We want to override that to move the highlight.

> there are quite a few places where we have buf_get_field()
> or even better do a trimmed down version like buf_get_field_callback()

Exactly.

In the test version of quick-search, the `buf_get_field()` function was ~600 lines long.
It had a gigantic switch statement for each of the possible functions.
They've all been farmed out into separate functions behind `enter_function_dispatcher()`

Of the remaining ~210 lines, some deal with history: we don't need that;
some deal with auto-completion: we don't need that either.

There's some GUI code that we might need to refactor, though.

Our personalised version of `buf_get_field()` would call the `sb_function_dispatcher()` first
allowing us to intercept <kbd>\<up\></kbd> and <kbd>\<down\></kbd>, then delegate the rest to `enter_function_dispatcher()`.

> I'm not entirely sure where we would use such quick input field in the near future

How about the Index?  or the Address Book?

> nor I fancy maintaining such a function over several refactoring jobs.

Once code is merged, you won't be tied to maintain it.
Of course I'd love you to stick around and implement more quick searches.
I think they'll be very popular.

> presenting 3 results with only the first one selected and
> What do you do then ? delete ? ^G it ? enter it ?

The user's narrowed down the list and used <kbd>\<up\></kbd> and <kbd>\<down\></kbd> to select one.
Then <kbd>\<enter\></kbd> will call `<sidebar-open>`
Hitting, say <kbd>\<esc\></kbd> or <kbd>\<Ctrl-G\></kbd> would abort
(hopefully leaving the screen looking exactly as it did before -- overlay :-)

> It might entirely be possible I'm overanalysing things

hehe, me too :-)

> It's really cheap to manipulate hil_index,...
> the cursor might jump a bit erratic but still workable I think.

Yes.  A simple (but slightly erratic) version first is a great start.

> a test dataset

I've just updated the sample mail repo with my "authors" list.
It contains 2956 mailboxes :-)

- https://github.com/neomutt/sample-mail

**Comment by lucilanga** [comment]
[![asciicast](https://asciinema.org/a/594259.svg)](https://asciinema.org/a/594259)

Not much of a recording I know, but I believe things are working. This is more of a porting The Matcher to the actual infra.
There is a problem with the overlay scrolling as you can see on Mr. Crowley or Blake, it comes from the paging I'm sure, but I haven't looked into that yet.

Should I just go ahead and open a PR for this and possibly get more feedback on this?
I do believe we've spammed this bug enough already.

Oh and thank you for the detailed info you provided earlier, it should have been easy enough.
It took me long enough but I believe this is time well spent as one needs to get accustomed to where things are, like where the headers are and what function does what.

~PS. I have no idea how to properly post this asciicast link~

**Comment by flatcap** [comment]
Excellent!

I've just sent you an invite to join NeoMutt.
I don't have an email address for you, so here's my [Welcome Email](https://gist.github.com/flatcap/b589998642e8a9b494e0a15af045ad29).
Feel free to drop me a line: rich@flatcap.org

---

> Should I just go ahead and open a PR for this and possibly get more feedback on this?

Yes please.
I suggest creating a branch on the `neomutt` repo, of the form `devel/SHORT-NAME`
That makes collaboration easier.

> thank you for the detailed info you provided

No problem.
NeoMutt's a lot of code.
Some bits are getting tidier, but there's still lots of twenty-year-old cruft lying about :-)

> get accustomed to where things are

We use Doxygen headers to document the code.
There are a couple of pages that may help you get an overview of NeoMutt:
- [NeoMutt Libraries](https://neomutt.org/code/pages.html) -- each subdir is a separate library
- [API Docs](https://neomutt.org/code/modules.html) -- Interfaces between blocks of code

Feel free to ask **LOTS** of questions; I'm always happy to help people help NeoMutt.
Ask here, in the PR, on IRC, or [open a discussion](https://github.com/neomutt/neomutt/discussions) if you have some big ideas

- IRC: `#neomutt` on `irc.libera.chat` ([web client](https://kiwiirc.com/nextclient/irc.libera.chat/#neomutt))

---

PS On the asciinema page, there's a **Share** link.  Click on that, then cut'n'paste the markdown.
